code/modules/projectiles/ammunition/ammo_pile.dm

```diff
@@ -0,0 +1,166 @@
+/obj/item/ammo_pile
+	name = "ammo pile"
+	desc = "A handful of some kind of ammunition."
+	w_class = ITEMSIZE_SMALL
+
+	var/list/ammo = list()
+	var/list/ammo_overlays = list()
+	var/image/ammo_overlay = list()
+	var/ammo_type // the type of ammo this ammo pile accepts
+	var/max_ammo = 5
+
+/obj/item/ammo_pile/Initialize(mapload, ...)
+	. = ..()
+	if(ammo_type)
+		var/obj/first_round = new ammo_type(get_turf(src))
+		add_ammo(first_round)
+		check_name_and_ammo()
+		for(var/i = 1, i <= max_ammo - 1, i++)
+			var/obj/C = new ammo_type(get_turf(src))
+			add_ammo(C)
+	addtimer(CALLBACK(src, .proc/check_ammo), 5) // if we don't have any ammo in 5 deciseconds, we're an empty pile, which is worthless, so self-delete
+
+/obj/item/ammo_pile/examine(mob/user)
+	. = ..()
+	if(Adjacent(user))
+		to_chat(user, SPAN_NOTICE("It contains [length(ammo)] rounds."))
+
+/obj/item/ammo_pile/attack()
+	return
+
+/obj/item/ammo_pile/afterattack(atom/A, mob/living/user, proximity_flag)
+	if(!proximity_flag)
+		return
+	if(!length(ammo))
+		check_ammo()
+
+	if(A.type == ammo_type)
+		if(length(ammo) >= max_ammo)
+			to_chat(user, SPAN_WARNING("\The [src] is already fully stacked."))
+			return
+		var/obj/item/ammo_casing/B = A
+		if(!B.BB)
+			to_chat(user, SPAN_WARNING("\The [B] is spent!"))
+			return
+		to_chat(user, SPAN_NOTICE("You add \the [A] to \the [src]."))
+		add_ammo(A)
+	else if(istype(A, /obj/item/ammo_pile))
+		if(length(ammo) >= max_ammo)
+			to_chat(user, SPAN_WARNING("\The [src] is already fully stacked."))
+			return
+		var/obj/item/ammo_pile/target_pile = A
+		if(target_pile.ammo_type != src.ammo_type)
+			to_chat(user, SPAN_WARNING("\The [target_pile] has a different type of ammunition!"))
+			return
+		var/amount_taken = 0
+		for(var/obj/bullet in target_pile.ammo)
+			if(length(src.ammo) >= src.max_ammo)
+				break
+			src.add_ammo(bullet)
+			target_pile.remove_ammo()
+			amount_taken++
+		to_chat(user, SPAN_NOTICE("You take [amount_taken] rounds from the other pile and add it to yours."))
+	else if(istype(A, /obj/item/gun) || istype(A, /obj/item/ammo_magazine))
+		var/obj/bullet = get_next_ammo()
+		A.attackby(bullet, user)
+		if(!(bullet in src)) // if the gun / mag acceoted the bullet, it will no longer be in our pile
+			remove_ammo()
+
+/obj/item/ammo_pile/attackby(obj/item/W, mob/user)
+	if(istype(W, /obj/item/ammo_casing))
+		if(W.type != ammo_type)
+			to_chat(user, SPAN_WARNING("\The [W] has a different type of ammunition!"))
+			return
+		if(length(ammo) >= max_ammo)
+			to_chat(user, SPAN_WARNING("\The [src] is already fully stacked."))
+			return
+		var/obj/item/ammo_casing/B = W
+		if(!B.BB)
+			to_chat(user, SPAN_WARNING("\The [B] is spent!"))
+			return
+		to_chat(user, SPAN_NOTICE("You add \the [W] to \the [src]."))
+		add_ammo(W)
```

Comment:
aren't all of these checks already done in add_ammo?

Reply:
sorta, the one in add_ammo is a safety measure to prevent spent casings from ever getting in. these have more checks that won't be earth shatteringly bad if they somehow were wrong